### ISSUE TITLE
Unexpected redirect to /auth when JWT has expired

### ISSUE THREAD

**Issue by optiz0r** [greenpau/caddy-auth-jwt#53]
Hi,

First up, thank you for the very useful plugins (both this, and caddy-auth-portal)!

I'm seeing unexpected behaviour around redirects after session expiration, not sure if it's a misconfiguration my side, or a bug.

- OK: When the user has no access_token cookie, a request to a protected resource issues a redirect to `/auth/auth2/generic?redirect_url=$protected`, which issues a redirect to the OIDC provider, and the user ends up back at `$protected`.
- OK: While the JWT is valid, the user ends up at `$protected` as expected.
- Bad: When `access_token` cookie contains an expired JWT, the user is directed to `/auth/oauth2/generic?redirect_url=$protected` as expected, but is then issued a redirect to `/auth` instead of the OIDC provider, leaving the user on the auth portal page instead of where they wanted to be.
- At this point, the cookie is gone, so reopening the protected URL results in the OIDC redirect shuffle and the user ends up back at the protected resource once again. 

In this setup, I never want the user to see the auth portal, and definitely don't want to leave them somewhere other than where they wanted to be when their session is expired, when the OIDC refresh should be seamless. I don't believe I have any config that should end up with the user on the /auth page.

**Configuration**

```
# simple app which echos back request headers
whoami.example.com {
  route {
    jwt {
      auth_url https://auth.example.com/auth/oauth2/generic
      default allow
    }
    reverse_proxy * {
      to "http://10.0.0.123:12345"
  }
}

https://auth.example.com {
  route /auth* {
    auth_portal {
      cookie_domain "example.com"
      backends {
        generic_oauth2_backend {
          method oauth2
          realm generic
          provider generic
          base_auth_url "https://adfs.example.com/adfs/oauth2/"
          metadata_url "https://adfs.example.com/adfs/.well-known/openid-configuration"
          client_id c49fcc02-2389-430f-b227-173ff1ee818a
          client_secret 776d2e35-1974-4ec8-be13-486efd4cf6cd
          scopes openid
        }
      }
      jwt {
        token_name access_token
        token_secret 64aea97d-984f-41d6-b293-1ecdc0f426b6
      }
    }
  }
  route /placeholder {
    jwt {
      primary yes
      auth_url /auth/oauth2/generic
      trusted_tokens {
        static_secret {
          token_name access_token
          token_secret 64aea97d-984f-41d6-b293-1ecdc0f426b6
        }
      }
      enable claim headers
      allow role any
    }
    respond "Nothing to see here, {http.request.header.X-Token-User-Name}!"
  } 
  route {
    # Avoid a blank white page, send the user somwhere useful
    redir /auth/whoami 302
  }
}
```

Domain/IPs obfuscated, and all secrets regenerated with random UUIDs
Provider is ADFS, but should not be relevant here.

**Version Information**

```
http.authentication.hashes.bcrypt v2.3.0
http.authentication.hashes.scrypt v2.3.0
http.authentication.providers.http_basic v2.3.0
http.authentication.providers.jwt v1.2.6
http.handlers.auth_portal v1.4.3
http.handlers.authentication v2.3.0
```

**Expected behavior**

`/auth/oauth2/generic?redirect_url=$protected` should issue a redirect to oidc provider instead of `/auth`

**Additional context**

Request details from chrome debug tools (redacted and trimmed for readability)

Previous request was `https://whoami.example.com` redirected to `https://auth.example.com/auth/oauth2/generic?redirect_url=https://whoami.example.com/`

```
General:
Request URL: https://auth.example.com/auth/oauth2/generic?redirect_url=https%3A%2F%2Fwhoami.example.com%2F
Request Method: GET
Status Code: 302 
Remote Address: 10.0.0.12:443
Referrer Policy: strict-origin-when-cross-origin

Response Headers:
content-length: 0
Content-Security-Policy-Report-Only: worker-src 'none'; report-uri about:blank
date: Mon, 08 Mar 2021 10:10:58 GMT
location: /auth
server: Caddy
set-cookie: AUTH_PORTAL_REDIRECT_URL=delete; Domain=example.com; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT;
set-cookie: access_token=delete; Domain=example.com; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT;

Request Headers:
:authority: auth.example.com
:method: GET
:path: /auth/oauth2/generic?redirect_url=https%3A%2F%2Fwhoami.example.com%2F
:scheme: https
accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
accept-encoding: gzip, deflate, br
accept-language: en-GB,en-US;q=0.9,en;q=0.8
cookie: access_token=eyREDACTED

Query String Parameters:
redirect_url: https://whoami.example.com/
```

**Comment by greenpau** [comment]
@optiz0r , thank you for opening the issue! This sounds like a bug. Will review and post back.

**Comment by optiz0r** [comment]
From reading through the code, I think this relates to `auth_url_path`. The redirect happens here: https://github.com/greenpau/caddy-auth-jwt/blob/2cac666bfebb1c74a0c074ef6dda0359b41a1316/pkg/auth/authorizer.go#L161

I found a brief reference to this in the README:

> The `auth_url_path` is the URL a user gets redirected to when a token is invalid.

Looks like this defaults to `/auth`
https://github.com/greenpau/caddy-auth-jwt/blob/2cac666bfebb1c74a0c074ef6dda0359b41a1316/pkg/auth/manager.go#L275
 and can be set via the json config, but not from the Caddyfile

```    
    jwt {
      primary yes
      auth_url /auth/oauth2/generic
      auth_url_path /auth/oauth2/generic
      trusted_tokens {
        static_secret {
          token_name access_token
          token_secret {env.JWT_SIGNING_SECRET}
        }
      }
      enable claim headers
      allow role any
    }
```

Produces error:

```
run: adapting config using caddyfile: parsing caddyfile tokens for 'route': /etc/caddy/Caddyfile:165 - Error during parsing: parsing caddyfile tokens for 'jwt': /etc/caddy/Caddyfile:154 - Error during parsing: unsupported root directive: auth_url_path
```

So, questions:
- Should an expired token be considered invalid token and redirect to /auth by default, or redirect to where `auth_url` points to try and recover/continue auth?
- Should `auth_url_path` be configurable from the Caddyfile?
- (What's the difference between `auth_url` and `auth_url_path`? there seems to be some overlap between what they do, but the names don't make it clear, at least to me, what the difference is)

**Comment by greenpau** [comment]
@optiz0r , I am guessing about this one.

Please read carefully this section: https://github.com/greenpau/caddy-auth-jwt#overview

Please ask questions if the above it unclear.

The following snippet is the "primary" for the "default" context.

```
  route /placeholder {
    jwt {
      primary yes
      auth_url https://auth.example.com/auth/oauth2/generic
```

The earlier `jwt` reference inherits its configuration from the above:

```
whoami.example.com {
  route {
    jwt {
      auth_url https://auth.example.com/auth/oauth2/generic
      default allow
    }
```

Enable debugging and see output for the jwt plugin. 

There could be a problem because both configure the same `jwt` context. However, I am not certain.

1. Remove `route /placeholder {` section completely and retry your tests.
2. Make the `whoami.example.com` primary for `jwt` or configure entirely new context.

**Comment by optiz0r** [comment]
I might be misunderstanding entirely.

> * There is only one primary instance in an authorization context.
> * A plugin MUST have a primary instance in an authorization context.

The way I'm using Caddy is with a container orchestrator. Jobs come and go at random, and may be opted into reverse proxying via labels. The orchestrator is writing out a caddyfile from a template, and triggering an online reload. At any given point in time, there may be zero, one or multiple sites being reverse proxied

Trying to make one (and only one) of the dynamically generated site blocks primary is overly complex and brittle. I can't make them all primary because of the above note in the docs. The placeholder route exists only so there is guaranteed to be one and only one `jwt { primary=yes}` instance in the config file (without this, sites wanting to use jwt will fail with error that no primary instance exists. It will never actually be used by anyone.

I see only these in the access log
```
{"level":"info","ts":1617004343.027656,"logger":"http.log.access.log2","msg":"handled request","request":{"remote_addr":"10.0.0.28:42498","proto":"HTTP/2.0","method":"GET","host":"whoami.example.com","uri":"/","headers":{"Sec-Fetch-Site":["none"],"Sec-Fetch-User":["?1"],"Accept-Encoding":["gzip, deflate, br"],"User-Agent":["blah"],"Sec-Ch-Ua-Mobile":["?0"],"Upgrade-Insecure-Requests":["1"],"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"],"Sec-Fetch-Mode":["navigate"],"Sec-Fetch-Dest":["document"],"Accept-Language":["en-GB,en-US;q=0.9,en;q=0.8"],"Cookie":["access_token=redacted"],"Sec-Ch-Ua":["\"Google Chrome\";v=\"89\", \"Chromium\";v=\"89\", \";Not A Brand\";v=\"99\""]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"h2","proto_mutual":true,"server_name":"whoami.example.com"}},"common_log":"10.0.0.28 - - [29/Mar/2021:07:52:23 +0000] \"GET / HTTP/2.0\" 302 12","duration":0.002315019,"size":12,"status":302,"resp_headers":{"Server":["Caddy"],"Set-Cookie":["access_token=delete; path=/; expires=Thu, 01 Jan 1970 00:00:00 GMT"],"Location":["https://auth.example.com/auth/oauth2/generic?redirect_url=https%3A%2F%2Fwhoami.example.com%2F"]}}

{"level":"info","ts":1617004343.0422492,"logger":"http.log.access.log6","msg":"handled request","request":{"remote_addr":"10.0.0.28:50064","proto":"HTTP/2.0","method":"GET","host":"auth.example.com","uri":"/auth/oauth2/generic?redirect_url=https%3A%2F%2Fwhoami.example.com%2F","headers":{"User-Agent":["blah"],"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"],"Sec-Fetch-Mode":["navigate"],"Sec-Fetch-Dest":["document"],"Sec-Ch-Ua-Mobile":["?0"],"Cookie":["access_token=redacted"],"Upgrade-Insecure-Requests":["1"],"Sec-Fetch-Site":["none"],"Sec-Fetch-User":["?1"],"Sec-Ch-Ua":["\"Google Chrome\";v=\"89\", \"Chromium\";v=\"89\", \";Not A Brand\";v=\"99\""],"Accept-Encoding":["gzip, deflate, br"],"Accept-Language":["en-GB,en-US;q=0.9,en;q=0.8"]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"h2","proto_mutual":true,"server_name":"auth.example.com"}},"common_log":"10.0.0.28 - - [29/Mar/2021:07:52:23 +0000] \"GET /auth/oauth2/generic?redirect_url=https%3A%2F%2Fwhoami.example.com%2F HTTP/2.0\" 302 0","duration":0.000513055,"size":0,"status":302,"resp_headers":{"Server":["Caddy"],"Set-Cookie":["AUTH_PORTAL_REDIRECT_URL=delete; Domain=example.com; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT;","access_token=delete; Domain=example.com; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT;"],"Location":["/auth"]}}

{"level":"info","ts":1617004343.0531564,"logger":"http.log.access.log6","msg":"handled request","request":{"remote_addr":"10.0.0.28:50064","proto":"HTTP/2.0","method":"GET","host":"auth.example.com","uri":"/auth","headers":{"Upgrade-Insecure-Requests":["1"],"Sec-Fetch-Mode":["navigate"],"Sec-Fetch-User":["?1"],"Sec-Ch-Ua-Mobile":["?0"],"Accept-Language":["en-GB,en-US;q=0.9,en;q=0.8"],"User-Agent":["blah"],"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"],"Sec-Fetch-Site":["none"],"Sec-Fetch-Dest":["document"],"Sec-Ch-Ua":["\"Google Chrome\";v=\"89\", \"Chromium\";v=\"89\", \";Not A Brand\";v=\"99\""],"Accept-Encoding":["gzip, deflate, br"]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"h2","proto_mutual":true,"server_name":"auth.example.com"}},"common_log":"10.0.0.28 - - [29/Mar/2021:07:52:23 +0000] \"GET /auth HTTP/2.0\" 200 1905","duration":0.000353186,"size":1905,"status":200,"resp_headers":{"Server":["Caddy"],"Cache-Control":["no-store"],"Pragma":["no-cache"],"Content-Type":["text/html"]}}
```

And this in the debug log
```
{"level":"error","ts":1617004343.0256743,"logger":"http.handlers.authentication","msg":"auth provider returned error","provider":"jwt","error":"[Token is expired]"}
```

Removing the placeholder block and making each site block primary results in Caddy not starting up with error:
```
{"level":"info","ts":1617006322.785383,"logger":"http.authentication.providers.jwt","msg":"provisioned plugin instance","instance_name":"jwt-1","started_at":1617006322.7852855}

run: loading initial config: loading new config: loading http app module: provision http: server srv1: setting up route handlers: route 5: loading handler modules: position 0: loading module 'subroute': provision http.handlers.subroute: setting up subroutes: route 0: loading handler modules: position 0: loading module 'subroute': provision http.handlers.subroute: setting up subroutes: route 0: loading handler modules: position 0: loading module 'authentication': provision http.handlers.authentication: loading authentication providers: module name 'jwt': provision http.authentication.providers.jwt: authentication provider registration error, instance jwt-2, error: found more than one primaryInstance instance of the plugin for default context
```

I see the plugin docs make multiple references to contexts and the default context, but no example or instructions for making multiple contexts. I had initially assumed only one was possible. If multiple contexts are possible, I need to understand how this is achieved.

**Comment by greenpau** [comment]
@optiz0r , after removing placeholder block, add “primary yes” to the block jwt under whoami and retest. Since you don’t have any primary left, the plugin throws an error, because there has to be a primary in a context.

**Comment by optiz0r** [comment]
whoami is not the only site block in the config file which should be using JWT, I just included that one to show the minimal reproduction case. The template looks like:

```
{{ range $vhost, $service := scratch.Get "vhosts" -}}
{{ $require_auth := eq (or (index $service.ServiceMeta "caddy_oidc") "false") "true" }}
{{ $vhost }} {
  route {
  {{- if $require_auth }}
    jwt {
      auth_url https://auth.infra.{{ env "BASE_DOMAIN" }}/auth/oauth2/generic
      {{- $auth_roles := index $service.ServiceMeta "caddy_require_groups" | split ","}}
      {{- if $auth_roles }}
      allow roles {{ range $auth_roles }}"{{ . }}"{{ end }}
      {{- else }}
      default allow
      {{- end }}
    }
  {{- end }}
  # ... other site config
{{- end }}
```

And there are multuple services, so the resulting config looks like:

```
whoami.example.com {
  route {
    jwt {
      auth_url https://auth.example.com/auth/oauth2/generic
      default allow
    }
}

foo.example.com {
  route {
    jwt {
      auth_url https://auth.example.com/auth/oauth2/generic
      default roles foobar
    }
}
```

If I set primary=yes inside this template block, then multiple sites get `primary=yes` and Caddy fails to start up with `error: found more than one primaryInstance instance of the plugin for default context`.

Since primary=yes is required exactly once in the config file, the placeholder block is that one place which is guaranteed to always be present; none of my services are guaranteed always be there, since they're controlled by the container orchestrator and updated in realtime.

Rather than have the placeholder be the primary for the "default context", ss there a way using Caddyfile to setup a separate "context" for each site, so that each one can set primary=yes, and duplicate the `trusted_tokens` block and be otherwise independent of each other?

**Comment by greenpau** [comment]
@optiz0r , add "context <unique name>“ and “primary yes” to each jwt section you create. Additionally, each primary context must have a reference to how to decode jwt tokens, i.e. the reverse of portal’s JWT configuration.

**Comment by optiz0r** [comment]
OK thanks - that bit is not clear from the README, there's no reference to a `context <string>` directive.

I've updated the config so that it looks like this:

```
whoami.example.com {
  route {
    jwt {
      primary yes
      context whoami.example.com
      auth_url https://auth.example.com/auth/oauth2/generic
      trusted_tokens {
        static_secret {
          token_name access_token
          token_secret {env.OIDC_RESIGNING_SECRET}
          token_lifetime 30
        }
      }
      enable claim headers
      default allow
    }
    # ...
  }
  log {
    output stdout
    level DEBUG
  }
}

foo.example.com {
  route {
    jwt {
      primary yes
      context foo.example.com
      auth_url https://auth.example.com/auth/oauth2/generic
      trusted_tokens {
        static_secret {
          token_name access_token
          token_secret {env.OIDC_RESIGNING_SECRET}
          token_lifetime 30
        }
      }
      enable claim headers
      allow roles "foo"
    }
    # ..
  }
  log {
    output stdout
    level DEBUG
  }
}
```

And reloaded Caddy. After waiting for the JWT in my `access_token` cookie to expire (which I've set down to 30s for testing), I reloaded the page, and got redirected to `/auth` again

access logs:
```
{"level":"info","ts":1617022605.7213712,"logger":"http.log.access.log6","msg":"handled request","request":{"remote_addr":"10.0.0.28:44520","proto":"HTTP/2.0","method":"GET","host":"auth.example.com","uri":"/auth/oauth2/generic?redirect_url=https%3A%2F%2Fwhoami.example.com%2F","headers":{"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36"],"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"],"Sec-Fetch-Mode":["navigate"],"Sec-Fetch-Dest":["document"],"Sec-Ch-Ua-Mobile":["?0"],"Accept-Language":["en-GB,en-US;q=0.9,en;q=0.8"],"Cookie":["access_token=redacted"],"Cache-Control":["max-age=0"],"Upgrade-Insecure-Requests":["1"],"Sec-Fetch-Site":["none"],"Sec-Fetch-User":["?1"],"Sec-Ch-Ua":["\"Google Chrome\";v=\"89\", \"Chromium\";v=\"89\", \";Not A Brand\";v=\"99\""],"Accept-Encoding":["gzip, deflate, br"]},"tls":{"resumed":true,"version":772,"cipher_suite":4865,"proto":"h2","proto_mutual":true,"server_name":"auth.example.com"}},"common_log":"10.0.0.28 - - [29/Mar/2021:12:56:45 +0000] \"GET /auth/oauth2/generic?redirect_url=https%3A%2F%2Fwhoami.example.com%2F HTTP/2.0\" 302 0","duration":0.001096682,"size":0,"status":302,"resp_headers":{"Server":["Caddy"],"Set-Cookie":["AUTH_PORTAL_REDIRECT_URL=delete; Domain=example.com; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT;","access_token=delete; Domain=example.com; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT;"],"Location":["/auth"]}}
{"level":"info","ts":1617022605.7294044,"logger":"http.log.access.log6","msg":"handled request","request":{"remote_addr":"10.0.0.28:44520","proto":"HTTP/2.0","method":"GET","host":"auth.example.com","uri":"/auth","headers":{"Cache-Control":["max-age=0"],"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"],"Sec-Fetch-Site":["none"],"Sec-Fetch-Mode":["navigate"],"Sec-Fetch-User":["?1"],"Sec-Fetch-Dest":["document"],"Sec-Ch-Ua":["\"Google Chrome\";v=\"89\", \"Chromium\";v=\"89\", \";Not A Brand\";v=\"99\""],"Sec-Ch-Ua-Mobile":["?0"],"Accept-Language":["en-GB,en-US;q=0.9,en;q=0.8"],"Upgrade-Insecure-Requests":["1"],"User-Agent":["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36"],"Accept-Encoding":["gzip, deflate, br"]},"tls":{"resumed":true,"version":772,"cipher_suite":4865,"proto":"h2","proto_mutual":true,"server_name":"auth.example.com"}},"common_log":"10.0.0.28 - - [29/Mar/2021:12:56:45 +0000] \"GET /auth HTTP/2.0\" 200 1905","duration":0.000765945,"size":1905,"status":200,"resp_headers":{"Server":["Caddy"],"Cache-Control":["no-store"],"Pragma":["no-cache"],"Content-Type":["text/html"]}}
```

error logs (grabbed from a different attempt so timestamps don't quite match up, but same behaviour observed both times)

```
{"level":"error","ts":1617027215.9788024,"logger":"http.handlers.authentication","msg":"auth provider returned error","provider":"jwt","error":"[Token is expired]"}
```

**Comment by greenpau** [comment]
@optiz0r , what is the callback URL configured in OIDC?

**Comment by optiz0r** [comment]
Callback URL is `https://auth.example.com/auth/oauth2/generic/authorization-code-callback`. The OIDC flow works fine when the user has no JWT, so I don't think there's anything wrong there. It's just that when the JWT is found to be expired, the user is directed to `/auth` instead of `/auth/oauth2/generic?redirect_url=$current_url` to restart the auto process.

**Comment by greenpau** [comment]
@optiz0r , this is a bug. I am going to address in the next couple of weeks.

what needs to happen here, is that when a portal receives redirect with an expired token, the portal should inspect the token and check whether “issuer” field. The field contains the callback URL. First, I would need to change the handling of how the field populates. Then, serve the appropriate landing page.

**Comment by greenpau** [comment]
FYI, started working on this one.

**Comment by greenpau** [comment]
First step is that I need to start moving away from `github.com/dgrijalva/jwt-go`.

Here, I would need to preserve the found invalid token somewhere and unpack the issuer from it.

**Comment by greenpau** [comment]
This is where the extracted token gets validated:

https://github.com/greenpau/caddy-auth-jwt/blob/e015954f483e26b032fa9af7b1e9a90faf57b361/pkg/validator/sources.go#L161-L169

In theory, we could add the logic of identifying issuer right here.

The next question is where do we put that URL. One choice is storing it in the `context.Context`.

Another one is adding a field to `User.Authenticator` struct.

Thinking it through 🤔 
r *http.Request

**Comment by optiz0r** [comment]
Thanks for getting to this!

I think this is working in so much as I got a response from my backend service without being left at a caddy-auth-portal view. However I am seeing some unexpected log entries in the Caddy stderr logs

```
{"level":"error","ts":1624886480.8840573,"logger":"http.handlers.authentication","msg":"auth provider returned error","provider":"jwt","error":"token validator: invalid token: keystore: failed to parse token"}
{"level":"error","ts":1624887861.216867,"logger":"http.handlers.authentication","msg":"auth provider returned error","provider":"jwt","error":"token validator: invalid token: keystore: failed to parse token"}
```

The first request was me hitting the whoami service for the first time since upgrading Caddy/caddy-auth-portal/caddy-auth-jwt. I then decoded the JWT cookie, waited until it's expiration time had passed, and refreshed the page. Dev tools confirmed three requests - Caddy for requested URL -> redirect to Caddy-auth-portal -> redirect to requested URL. Surprisingly I didn't see a redirect to the OIDC provider, but did end up at the backend service as an authenticated user.

Not sure if the logs above are indicative of a problem, or safe to ignore.
